homeassistant/components/switch/tellduslive.py

```diff
@@ -21,54 +20,49 @@ def setup_platform(hass, config, add_devices, discovery_info=None):
     """ Find and return Tellstick switches. """
     if discovery_info is None:
         return
-    switches = tellduslive.NETWORK.get_switches()
-    add_devices([TelldusLiveSwitch(switch["name"],
-                                   switch["id"])
-                 for switch in switches if switch["type"] == "device"])
+    add_devices(TelldusLiveSwitch(switch) for switch in discovery_info)
 
 
 class TelldusLiveSwitch(ToggleEntity):
     """ Represents a Tellstick switch. """
 
-    def __init__(self, name, switch_id):
-        self._name = name
+    def __init__(self, switch_id):
         self._id = switch_id
-        self._state = STATE_UNKNOWN
         self.update()
+        _LOGGER.debug("created switch %s", self)
+
+    def update(self):
+        tellduslive.NETWORK.update_switches()
+        self._switch = tellduslive.NETWORK.get_switch(self._id)
 
     @property
     def should_poll(self):
         """ Tells Home Assistant to poll this entity. """
-        return False
+        return True
+
+    @property
+    def _switch_state(self):
+        return self._switch["state"]
 
     @property
     def name(self):
         """ Returns the name of the switch if any. """
-        return self._name
-
-    def update(self):
-        from tellive.live import const
-        state = tellduslive.NETWORK.get_switch_state(self._id)
-        if state == const.TELLSTICK_TURNON:
-            self._state = STATE_ON
-        elif state == const.TELLSTICK_TURNOFF:
-            self._state = STATE_OFF
-        else:
-            self._state = STATE_UNKNOWN
+        return self._switch["name"]
 
     @property
     def is_on(self):
         """ True if switch is on. """
-        return self._state == STATE_ON
+        from tellive.live import const
+        return self._switch_state == const.TELLSTICK_TURNON
 
     def turn_on(self, **kwargs):
         """ Turns the switch on. """
         if tellduslive.NETWORK.turn_switch_on(self._id):
-            self._state = STATE_ON
-            self.update_ha_state()
+            from tellive.live import const
```

Comment:
You are mutating data here from the `NETWORK` object since the dicts are linked between the entity and the `NETWORK` object. Because of that, I would expect the optimistic state update logic to live in the `NETWORK` object.